### PR TITLE
Simplify SpectralCoord implementation

### PR DESCRIPTION
**Pull request by astrofrog**
The current SpectralCoord implementation relies on an internal 'fake' observer and target when these are not specified explicitly, but I believe this actually makes the implementation more complicated than it needs to be (the original intent was to actually make it simpler but this is not the case in the end!).

In addition, some of the implementation is also overly complex because of an original use case (changing the actual observer location and taking into account parallax effects) which didn't make it into the final API since we determined that this might not be needed. This PR shows how we can simplify the implementation while still having the tests pass (at least they do for me locally).

Even if we do eventually implement more complex use cases, I think we should keep the implementation as simple as possible for the current use cases, for maintainability and speed.

The resulting code in this PR has fewer frame transformations and vector operations, and is therefore faster. For instance:

**Before**

```python
In [1]: from astropy.coordinates import SpectralCoord                                                                                                                            

In [2]: from astropy import units as u                                                                                                                                           

In [3]: velocity = 100 * u.km / u.s                                                                                                                                              

In [4]: %timeit sc = SpectralCoord(100, 'GHz')                                                                                                                                   
2.77 ms ± 121 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [5]: sc = SpectralCoord(100, 'GHz')                                                                                                                                   

In [6]: %timeit res = sc.with_radial_velocity_shift(velocity)                                                                                                                    
7.11 ms ± 410 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**After**

```python
In [1]: from astropy.coordinates import SpectralCoord                                                                                                                            
^[[A^[[A
In [2]: from astropy import units as u                                                                                                                                           

In [3]: velocity = 100 * u.km / u.s                                                                                                                                              

In [4]: %timeit sc = SpectralCoord(100, 'GHz')                                                                                                                                   
134 µs ± 2.37 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: sc = SpectralCoord(100, 'GHz')                                                                                                                                           

In [6]: %timeit res = sc.with_radial_velocity_shift(velocity)                                                                                                                    
418 µs ± 34.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

So the initialization in this case is 20x faster and the call to ``with_radial_velocity_shift`` is 17x faster. This is just an example of course, but overall I think all operations will be the same or faster.

This will require a careful review to make sure I haven't actually broken anything. It might be possible to simplify things further but I haven't had a chance to work on this more as I'm not going to be working much on astropy for a couple of months. So if anyone would like to review this in detail and push commits to it, please feel free to do so!

(cc @mhvk who originally commented on the SpectralCoord PR that the internal coordinates seemed unnecessary)

### CODE REVIEW

File: astropy/wcs/wcsapi/sliced_low_level_wcs.py
Comment (Cadair) [inline astropy/wcs/wcsapi/sliced_low_level_wcs.py:6]: why you take away my warning? :sob:

Comment (astrofrog) [reply astropy/wcs/wcsapi/sliced_low_level_wcs.py:6]: Oops sorry :)

Comment (astrofrog) [reply astropy/wcs/wcsapi/sliced_low_level_wcs.py]: Fixed

File: astropy/coordinates/spectral_coordinate.py
Comment (mhvk) [inline astropy/coordinates/spectral_coordinate.py:464]: Is it good to have this as a default? Or would `AttributeError` be more reasonable? Though I can see that if one wants there to be a (reasonable) default of 0, it is good to keep track of the fact that it is entered as a default, not an explicit value.

File: astropy/coordinates/spectral_coordinate.py
Comment (mhvk) [inline astropy/coordinates/spectral_coordinate.py:607]: This makes me wonder, why not allow the velocity to be specified as a `CartesianDifferential` in the first place? With a recent PR of mine, that will now be passed along, but you have to *not* expand the velocity (which is fine, a triplet can be used as input anyway). Anyway, bottom line, would suggest to remove the `*` here and update the docstring that a `CartesianDifferential` is allowed...

Comment (mhvk) [reply astropy/coordinates/spectral_coordinate.py:607]: And then below too work in terms of a differential... (since this is really rather off-topic, do feel free to ignore).

File: astropy/coordinates/tests/test_spectral_coordinate.py
Comment (mhvk) [inline astropy/coordinates/tests/test_spectral_coordinate.py:526]: Probably should check what `radial_velocity` gives here too.

File: astropy/coordinates/spectral_coordinate.py
Comment (nmearl) [inline astropy/coordinates/spectral_coordinate.py:675]: Is there a particular reason we still replicate instead of just returning `self`?

File: astropy/coordinates/spectral_coordinate.py
Comment (nmearl) [inline astropy/coordinates/spectral_coordinate.py:720]:
```suggestion

```
No longer relevant, really.

File: astropy/coordinates/tests/test_spectral_coordinate.py
Comment (nmearl) [reply astropy/coordinates/tests/test_spectral_coordinate.py:526]: Isn't that done in the line below?

File: astropy/coordinates/spectral_coordinate.py
Comment (pllim) [reply astropy/coordinates/spectral_coordinate.py:675]: Re: just returning self -- Then it won't be a copy anymore, like the docstring claims?

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:22]: ni

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:283]: See https://github.com/astropy/astropy/issues/10247 - I'll add a link to the issue in the comment above

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:464]: Yes the idea was to differentiate between explicitly specified zero versus default though maybe that distinction is unimportant. I don't feel strongly about this, but I think raising an ``AttributeError`` would be too much.

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:22]: Fixed :)

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:27]: Fixed

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:49]: Removed

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:141]: Fixed

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:289]: Fixed

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:356]: Agree, fixed!

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:364]: Done

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:267]: Done

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:419]: Done

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:445]: Done

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:607]: I've adjusted it to allow ``CartesianDifferential`` to be passed in

File: astropy/coordinates/tests/test_spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/tests/test_spectral_coordinate.py:526]: Indeed this is done below

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:675]: Yep - I would always expect replicate to return a copy

**Comment by astrofrog** [comment]
@mhvk - thanks for the detailed review! I think I've addressed all your comments

@nmearl - thanks for the suggestions! I've included them here.

File: astropy/coordinates/spectral_coordinate.py
Comment (nmearl) [reply astropy/coordinates/spectral_coordinate.py:678]: Right, of course -- `replicate` _should_ return a copy. My main question is: why are we replicating in this case? If a user gets here, they're not quite using the method as intended _and_ they're duplicating data which gives the impression that ✨ something happened ✨, but really we internally deal with this as "no observer or target; can't do anything; returning". Perhaps a warning to make the user aware, or an error to make it clear that this method _does_ require arguments to work properly?

Overall, a minor quibble, so no need to address this with any earnest. 😁

Comment (pllim) [reply astropy/coordinates/spectral_coordinate.py:678]: If they not supposed to get to that part of the code, then throwing an exception would be the safest?

**Comment by astrofrog** [comment]
I've rebased this - I'd really like to get this in v4.2 if possible.

@mhvk @nmearl - are you happy with this now?

Anyone else want to review this? (@eteq?)

File: astropy/coordinates/spectral_coordinate.py
Comment (mhvk) [reply astropy/coordinates/spectral_coordinate.py:678]: @astrofrog - no conclusion seems to have been reached here...

**Comment by mhvk** [comment]
@astrofrog - I think there is just one outstanding comment... Otherwise looks good!

File: astropy/coordinates/spectral_coordinate.py
Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:678]: There could be programmatic reasons why this method gets called with no arguments (e.g if iterating over many objects and using some logic that means that sometimes it is called with no arguments) - in this case I would always want the results to be new objects, not self sometimes and a new object other times. But maybe I can make this explicit in the docstring.

Comment (astrofrog) [reply astropy/coordinates/spectral_coordinate.py:679]: See 3465d53abe7b66922cc55848d91419a3457e12fb for a clarification in the docstring. Is this ok now?

Comment (mhvk) [reply astropy/coordinates/spectral_coordinate.py:679]: Looks good to me

**Review by mhvk — APPROVED** [review]
I think it all looks good to me - agreed this is a nice simplification that should find its way in!

**Comment by astrofrog** [comment]
Ok thanks! I'll look into fixing the CI today.

**Review by Cadair — APPROVED** [review]
Well it didn't break the ndcube tests...

**Comment by Cadair** [comment]
@astrofrog I rebased to fix the merge conflict and fixed the flake8 issue :smile:

**Comment by astrofrog** [comment]
@StuartLittlefair @adrn @astrojuanlu @eteq @mhvk - it would be best for this to be merged by a coordinates maintainer, so I'll leave it up to you to merge. I think it'd be a shame to miss v4.2 though so please merge today if possible! 🙏

**Comment by StuartLittlefair** [comment]
Hi @astrofrog - since @mhvk has run his fine tooth comb over this PR I'm happy to merge in principle, but I notice from the coverage that the cases where, for example, ```self._target``` and ```self._observer``` are both None are not being tested. There are a number of these lines. 

Maybe worth looking to see if the code is well tested in these instances before hitting the button?

**Comment by astrofrog** [comment]
@StuartLittlefair - I've tried to improve the test coverage, now 97% for the ``spectral_coordinate.py`` file. I won't have time to get to 100% today but hopefully this should be good enough?

**Comment by pllim** [comment]
Tsk tsk... Should have checked the remote data job. Can you please fix the failing doctest?

```
_________________________ [doctest] spectralcoord.rst __________________________
310 method. This method can take the name of an existing coordinate/velocity frame,
311 a :class:`~astropy.coordinates.BaseCoordinateFrame` instance, or any arbitrary
312 3D position and velocity coordinate object defined either as a
313 :class:`~astropy.coordinates.BaseCoordinateFrame` or a |SkyCoord| object. Most
314 commonly-used frames are accessible using strings. For example to transform to a
315 velocity frame stationary with respect to the center of the Earth (so removing
316 the effect of the Earth's rotation), we can use the ``'gcrs'`` which stands for
317 *Geocentric Celestial Reference System* (GCRS)::
318 
319     >>> sc_ttau.with_observer_stationary_relative_to('gcrs')  # doctest: +FLOAT_CMP +REMOTE_DATA
Differences (unified diff with -expected +actual):
    @@ -1,7 +1,7 @@
    -<SpectralCoord
    +<SpectralCoord
        (observer: <GCRS Coordinate (obstime=2019-04-24T02:32:10.000, obsgeoloc=(0., 0., 0.) m, obsgeovel=(0., 0., 0.) m / s): (x, y, z) in m
    -                  (-5878853.86171412, -192921.84773269, -2470794.19765021)
    +                  (-5878853.86172938, -192921.84773269, -2470794.19765021)
                    (v_x, v_y, v_z) in km / s
    -                  (4.33251262e-09, 8.96175625e-08, -1.49258412e-08)>
    +                  (-1.09244138e-08, 8.96161946e-08, -1.49267726e-08)>
         target: <ICRS Coordinate: (ra, dec, distance) in (deg, deg, pc)
                     (65.497625, 19.53511111, 144.321)
    @@ -9,6 +9,6 @@
                     (1.37949782e-15, 1.46375638e-15, 23.9)>
         observer to target (computed from above):
    -      radial_velocity=40.674086368345165 km / s
    -      redshift=0.00013568335316072044)
    +      radial_velocity=40.67408630725199 km / s
    +      redshift=0.0001356833529568835)
       [200.00024141, 210.00025348, 220.00026555, 230.00027762, 240.00028969,
        250.00030176, 260.00031383, 270.0003259 , 280.00033797, 290.00035004,
/home/travis/build/astropy/astropy/docs/coordinates/spectralcoord.rst:319: DocTestFailure
```

https://travis-ci.com/github/astropy/astropy/jobs/404743171 (This PR)

https://travis-ci.com/github/astropy/astropy/jobs/404850035 (On `master` after merge)

**Comment by astrofrog** [comment]
Huh wait but I did check it locally and that's what led to the doctest updates in https://github.com/astropy/astropy/pull/10398/commits/26442d19cdfc87f3492f21cc638694145f525b03 - I wonder why the results are different on Travis?!

**Comment by pllim** [comment]
> I wonder why the results are different on Travis?!

The same job also uses numpy dev, though not sure if that is the cause or not. If only we can use https://docs.travis-ci.com/user/running-build-in-debug-mode/

**Comment by StuartLittlefair** [comment]
argh. I'm away from the computer until late tonight. Sorry for the premature merge!

**Comment by mhvk** [comment]
Looks like some rounding error problem - these velocities are equal to 0 km/s within a few μm/s... Just skip the test until we have a `FLOAT_CMP` where we also can give absolute tolerances?

**Comment by pllim** [comment]
Started attempt at #10928

**Pull request merged**
